source/common/network/listen_socket_impl.h

```diff
@@ -244,14 +244,25 @@ class ConnectionSocketImpl : public SocketImpl, public ConnectionSocket {
 class AcceptedSocketImpl : public ConnectionSocketImpl {
 public:
   AcceptedSocketImpl(IoHandlePtr&& io_handle, const Address::InstanceConstSharedPtr& local_address,
-                     const Address::InstanceConstSharedPtr& remote_address)
-      : ConnectionSocketImpl(std::move(io_handle), local_address, remote_address) {
-    ++global_accepted_socket_count_;
+                     const Address::InstanceConstSharedPtr& remote_address,
+                     Server::ThreadLocalOverloadStateOptRef overload_state,
+                     bool track_global_cx_limit_in_overload_manager)
+      : ConnectionSocketImpl(std::move(io_handle), local_address, remote_address),
+        overload_state_(overload_state),
+        track_global_cx_limit_in_overload_manager_(track_global_cx_limit_in_overload_manager) {
+    if (!track_global_cx_limit_in_overload_manager_) {
+      ++global_accepted_socket_count_;
+    }
   }
 
   ~AcceptedSocketImpl() override {
-    ASSERT(global_accepted_socket_count_.load() > 0);
-    --global_accepted_socket_count_;
+    if (track_global_cx_limit_in_overload_manager_) {
+      overload_state_->tryDeallocateResource(
+          Server::OverloadProactiveResourceName::GlobalDownstreamMaxConnections, 1);
+    } else {
+      ASSERT(global_accepted_socket_count_.load() > 0);
```

Comment:
Hm. I'm slightly concerned about this going away. I've hit this ASSERT() in the past by buggingly changing some socket accounting. Do we need to remove this assert, even if we're using the overload manager generally?

Reply:
@RyanTheOptimist we could add similar assert in overload manager `if branch` above as extra layer of protection.

Reply:
I think that would be great.

Reply:
while this protection would be useful for programmatic errors, it will not cover all race cases, since check for `current.load() > 0` reads atomic value in one place and then later tries to decrement it, there is still a chance of race condition (e.g. another worker threads have consumed all resource usage between assert and actual decrement). Overload manager logic has protection against this case: https://github.com/envoyproxy/envoy/blob/main/source/extensions/resource_monitors/downstream_connections/downstream_connections_monitor.cc#L41

Reply:
@RyanTheOptimist do you think assertion is still needed given that overload manager method would fail to deallocate resource if current value is <=0 ?

Reply:
Ah, I see. Ok, fair enough.